feat(youtube): add support for new endpoints

LuizDMM opened the pull request:
Fixes #9
/claim #9

**Requirements** 

- [x] Support for above endpoints according to the documentation and aligned with patterns in this library is added
- [x] All parameters and returns are typed according to spec
- [x] README is updated
- [x] Test cases are added in [tests](https://github.com/supadata-ai/py/blob/main/tests/test_client.py)

Implemented the support for all required endpoints, in a pythonic and typed way.  Implemented tests using the content from some real requests I made and all the tests are passing.

```
========================================================================================= test session starts ==========================================================================================
platform linux -- Python 3.12.9, pytest-8.3.5, pluggy-1.5.0
rootdir: /home/luizdmm/Projetos/Algora Bounties/supadata-ai-py
configfile: pyproject.toml
plugins: requests-mock-1.12.1
collected 21 items                                                                                                                                                                                     

tests/test_client.py .....................                                                                                                                                                       [100%]

========================================================================================== 21 passed in 0.08s ==========================================================================================
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced YouTube integration to let users retrieve comprehensive metadata for videos, channels, and playlists.
  - Introduced functionality for fetching lists of video identifiers from channels and playlists, providing a richer view of YouTube content.
  - Improved output messages to display retrieved information clearly, offering an optimized experience when accessing YouTube data.

- **Bug Fixes**
  - Updated error handling for API requests to ensure accurate status codes and messages.

- **Tests**
  - Expanded test coverage for YouTube functionalities, including new tests for video, channel, and playlist retrieval.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LuizDMM commented on supadata/youtube.py:
```diff
@@ -84,8 +91,118 @@ def translate(
+            "GET", "/youtube/playlist/videos", params=query_params
+        )
+
+        return response.get("video_ids", [])
```
The method `Supadata._camel_to_snake` converts the string `"VideoIds"` to `"video_ids"`, that said the method works and a response with elements is covered in the test cases.

LuizDMM commented on supadata/youtube.py:
```diff
@@ -84,8 +91,118 @@ def translate(
+            "GET", "/youtube/channel/videos", params=query_params
+        )
+
+        return response.get("video_ids", [])
```
The method `Supadata._camel_to_snake` converts the string `"VideoIds"` to `"video_ids"`, that said the method works and a response with elements is covered in the test cases.

rafalzawadzki commented on tests/test_client.py:
```diff
@@ -249,66 +211,295 @@ def test_get_crawl_results(client: Supadata, requests_mock) -> None:
+    mock_response = {
+        "error": "not-found",
+        "message": "The requested item could not be found",
+        "details": "The requested item could not be found.",
+    }
+
+    requests_mock.get(
+        f"{client.base_url}/youtube/playlist/videos?id={playlist_id}",
+        status_code=404,
+        json=mock_response,
+    )
+
+    with pytest.raises(SupadataError) as exc_info:
+        client.youtube.playlist_videos(playlist_id)
+
     error = exc_info.value
-    assert error.error == 'limit-exceeded'
-    assert error.message == 'Limit exceeded'
-    assert error.details == 'You have exceeded the allowed request rate or quota limits' 
+    assert error.error == "invalid-request"
+    assert error.message == "Endpoint does not exist"
+    assert error.details == "The requested item could not be found"
```
is the assertion correct?

the error and message seem to be different than the mock response

LuizDMM replied:
Yes, as explained in another one, due to `Supadata._handle_gateway_error`.

rafalzawadzki commented on tests/test_client.py:
```diff
@@ -249,66 +211,295 @@ def test_get_crawl_results(client: Supadata, requests_mock) -> None:
+
+def test_youtube_channel_videos_invalid_id(client: Supadata, requests_mock) -> None:
+    channel_id = "UCsBjURrPoezyLs9EqgamOA"
+    mock_response = {
+        "error": "not-found",
+        "message": "The requested item could not be found",
+        "details": "The requested item could not be found.",
+    }
+    requests_mock.get(
+        f"{client.base_url}/youtube/channel/videos?id={channel_id}",
+        status_code=404,
+        json=mock_response,
+    )
+
+    with pytest.raises(SupadataError) as exc_info:
+        client.youtube.channel_videos(channel_id)
+
+    error = exc_info.value
+    assert error.error == "invalid-request"
+    assert error.message == "Endpoint does not exist"
+    assert error.details == "The requested item could not be found"
```
is the assertion correct?

the error and message seem to be different than the mock response

LuizDMM replied:
Yes, as explained in another one, due to `Supadata._handle_gateway_error`.

rafalzawadzki commented on tests/test_client.py:
```diff
@@ -249,66 +211,295 @@ def test_get_crawl_results(client: Supadata, requests_mock) -> None:
+
+def test_youtube_playlist_invalid_id(client: Supadata, requests_mock) -> None:
+    playlist_id = "PL0vfts4VzfNjQOM9VClyL50LeuTxlAR3"
+    mock_response = {
+        "error": "not-found",
+        "message": "The requested item could not be found",
+        "details": "The requested item could not be found.",
+    }
+
+    requests_mock.get(
+        f"{client.base_url}/youtube/playlist?id={playlist_id}",
+        status_code=404,
+        json=mock_response,
+    )
+
+    with pytest.raises(SupadataError) as exc_info:
+        client.youtube.playlist(playlist_id)
+
+    error = exc_info.value
+    assert error.error == "invalid-request"
+    assert error.message == "Endpoint does not exist"
+    assert error.details == "The requested item could not be found"
```
is the assertion correct?

the error and message seem to be different than the mock response

rafalzawadzki commented on tests/test_client.py:
```diff
@@ -249,66 +211,295 @@ def test_get_crawl_results(client: Supadata, requests_mock) -> None:
+
+
+def test_youtube_channel_invalid_id(client: Supadata, requests_mock) -> None:
+    channel_id = "UCsBjURrPoezyLs9EqgamOA"
+    mock_response = {
+        "error": "not-found",
+        "message": "The requested item could not be found",
+        "details": "The requested item could not be found.",
+    }
+    requests_mock.get(
+        f"{client.base_url}/youtube/channel?id={channel_id}",
+        status_code=404,
+        json=mock_response,
+    )
+    with pytest.raises(SupadataError) as exc_info:
+        client.youtube.channel(channel_id)
+
+    error = exc_info.value
+    assert error.error == "invalid-request"
+    assert error.message == "Endpoint does not exist"
+    assert error.details == "The requested item could not be found"
```
is the assertion correct?

the error and message seem to be different than the mock response

LuizDMM replied:
Yes, the assertion is correct. The difference between the message from the API and the one in the exception is due to `Supadata._handle_gateway_error`. When the API returns 404, the message is always "Endpoint does not exist.", what changes is the details, that is based from the message in the response:

![image](https://github.com/user-attachments/assets/1d71d16d-3795-46a5-b115-17daea10c54e)

rafalzawadzki commented on tests/test_client.py:
```diff
@@ -249,66 +211,295 @@ def test_get_crawl_results(client: Supadata, requests_mock) -> None:
+
+def test_youtube_video_invalid_id(client: Supadata, requests_mock) -> None:
+    video_id = "pEfrdAtmqk"
+    mock_response = {
+        "error": "not-found",
+        "message": "The requested item could not be found",
+    }
+    requests_mock.get(
+        f"{client.base_url}/youtube/video?id={video_id}",
+        status_code=404,
+        json=mock_response,
+    )
+
+    with pytest.raises(SupadataError) as exc_info:
+        client.youtube.video(video_id)
+
+    error = exc_info.value
+    assert error.error == "invalid-request"
+    assert error.message == "Endpoint does not exist"
+    assert error.details == "The requested item could not be found"
```
is the assertion correct?

the error and message seem to be different than the mock response

rafalzawadzki reviewed (changes requested):
hi @LuizDMM thanks for contributing your solution!

mostly looks good, but I have two asks:
1. Asserts seem to be different from mocks in the tests. Are they correct?
2. Can you add namespaces for channel and playlist so we can accomplish dot notation like with other features? eg. `channel.videos` instead of `channel_videos`

LuizDMM commented:
> hi @LuizDMM thanks for contributing your solution!
> 
> mostly looks good, but I have two asks:
> 
>     1. Asserts seem to be different from mocks in the tests. Are they correct?
> 
>     2. Can you add namespaces for channel and playlist so we can accomplish dot notation like with other features? eg. `channel.videos` instead of `channel_videos`

Hey @rafalzawadzki, thank you for your feedback! I'll check all the tests and the namespace and get back to you.

LuizDMM commented:
Hey @rafalzawadzki, done! Just some comments about what I've done and your questions:

> 1. Asserts seem to be different from mocks in the tests. Are they correct?

Just repeating myself from above, yes, the assertion is correct. The difference between the message from the API and the one in the exception is due to `Supadata._handle_gateway_error`. When the API returns 404, in the case of a 404, `SupadataError.message` is always _"Endpoint does not exist."_, what changes is the `SupadataError.details`, that is based from the message in the response as you can see in the code [here](https://github.com/supadata-ai/py/blob/main/supadata/client.py#L34) (`supadata/client.py `lines 58-63):

![Captura de tela de 2025-03-18 20-49-39](https://github.com/user-attachments/assets/3c774d99-eca5-4c60-8fff-66502ba0271d)

We can take a look at this and modify `Supadata._handle_gateway_error` if you wish. Just let me know and I'll fix this with another commit.

> 2. Can you add namespaces for channel and playlist so we can accomplish dot notation like with other features? eg. channel.videos instead of channel_videos

Sure can, I added two inner classes to `YouTube`, `_Channel` and `_Playlist`, moved the logic to these classes, so now you can pull the data in the desired dot notation. I also updated the documentation and tests to reflect these changes.


I hope these changes fix everything, but just let me know if you want me to change the handler of gateway error or anything else in the PR

rafalzawadzki commented:
oo good find! this is a bug - the gateway should return `not-found` for 404 errors instead of `invalid-request`.

Can you please fix that? I'll add a bonus to this bounty to compensate your extra effort.

LuizDMM commented:
> oo good find! this is a bug - the gateway should return `not-found` for 404 errors instead of `invalid-request`.
> 
> Can you please fix that? I'll add a bonus to this bounty to compensate your extra effort.

Sure, on it.

LuizDMM commented:
Done! Fixed the message and the tests to reflect the change.

rafalzawadzki reviewed (approved):
thanks @LuizDMM! lgtm 🚀

rafalzawadzki commented:
/tip $10
